### PR TITLE
BPU Noop insertion pass

### LINKED ISSUE #4: Revive no-op insertion pass for RISC-V
You'll need to revive the reverted [random `nop`s insertion pass](https://reviews.llvm.org/D3392). 
High-level steps
- Read the review
- Reproduce the patch
- Transfer the patch to RISC-V
- Verify the pass actually inserts `nop`s by using [this](https://stackoverflow.com/questions/11227809/why-is-processing-a-sorted-array-faster-than-processing-an-unsorted-array?m=1) test (`objdump -d` for RISC-V target).

### PR DESCRIPTION
**Pull request by d-tolmachev**
Randomly insert nop-instructions via `fdiversify` option

### CODE REVIEW

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 1]
```diff
@@ -9349,6 +9350,52 @@ void X86InstrInfo::insertNoop(MachineBasicBlock &MBB,
   BuildMI(MBB, MI, DL, get(X86::NOOP));
 }
 
+/// insertNoop - Insert a randomly chosen type of noop into the instruction
+/// stream at the specified point to introduce fine-grained diversity.
+void X86InstrInfo::insertNoop(MachineBasicBlock &MBB,
```
Comment (ArsenyBochkarev) [inline llvm/lib/Target/X86/X86InstrInfo.cpp]: I guess we don't need X86 (for now, at least :grin:)

File: llvm/test/CodeGen/X86/noop-insert-percentage.ll [hunk 2]
```diff
@@ -0,0 +1,66 @@
+; RUN: llc < %s -mtriple=x86_64-linux -rng-seed=5 -noop-insertion -noop-insertion-percentage=10 \
```
Comment (ArsenyBochkarev) [inline llvm/test/CodeGen/X86/noop-insert-percentage.ll]: I'm fine with tests but for RISC-V. Let's remove it just for consistency

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -6960,6 +6960,17 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
...
                      options::OPT_mno_stack_arg_probe);
 
+  // Translate -frandom-seed to seed the LLVM RNG
+  if (Args.hasArg(options::OPT_frandom_seed_EQ)) {
+    StringRef seed = Args.getLastArgValue(options::OPT_frandom_seed_EQ);
+    CmdArgs.push_back("-mllvm");
+    CmdArgs.push_back(Args.MakeArgString("-rng-seed=" + seed));
+  }
+
+  if (Args.hasArg(options::OPT_fdiversify)) {
```
Comment (ArsenyBochkarev) [inline clang/lib/Driver/ToolChains/Clang.cpp:6971]: Could you please *also* push these options when `--fuzz=bpu` flag is passed to clang?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 4]
```diff
@@ -8081,10 +8092,13 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
...
         "all", "scheduler", "mbb-placement", "regalloc", "isel", "alloca", "bpu"};
 
-    for (const auto &ActualOpt : ActualOpts)
+    for (const auto &ActualOpt : ActualOpts) {
       if (llvm::find(CorrectOpts, ActualOpt) == CorrectOpts.end())
         D.Diag(diag::err_analyzer_checker_option_unknown)
             << "fuzzer" << ActualOpt;
+      if (ActualOpt == "all" || ActualOpt == "bpu")
+        CmdArgs.push_back("-noop-insertion");
+    }
```
Comment (ArsenyBochkarev) [inline clang/lib/Driver/ToolChains/Clang.cpp]: Please note the [current state](https://github.com/Compiler-assisted-fuzzing/llvm-project/blob/dev/clang/lib/Driver/ToolChains/Clang.cpp#L8097) of this loop

**Comment by d-tolmachev** [comment]
Squashed fixes into single commit and merge with dev branch

**Pull request merged**